spec/support/authentication_service_shared_examples.rb

```diff
@@ -0,0 +1,452 @@
+          ).where(
+            'comment @> ?', {action: called_action, endpoint: url}.to_json
+          ).order(:created_at).last
+          expect(last_audit.user).to be
+          expect(last_audit.user.username).to eq(first_time_user[:username])
+        end
+      end
+    end
+
+    describe 'for existing users' do
+      let(:payload) {
+        if authentication_service.is_default?
+          {
+            access_token: existing_user_access_token
+          }
+        else
+          {
+            access_token: existing_user_access_token,
+            authentication_service_id: authentication_service.service_id
+          }
+        end
+      }
+
+      it 'should update user.last_login_at and return an api JWT when provided a JWT access_token encoded with our secret by a registered AuthenticationService' do
+        original_last_login_at = existing_user.last_login_at.to_i
+        pre_time = DateTime.now.to_i
+        is_expected.to eq(200)
+        post_time = DateTime.now.to_i
+        expect(response.status).to eq(200)
+        expect(response.body).to be
+        expect(response.body).not_to eq('null')
+        token_wrapper = JSON.parse(response.body)
+        expect(token_wrapper).to have_key('expires_on')
+        expect(token_wrapper).to have_key('api_token')
+        decoded_token = JWT.decode(token_wrapper['api_token'],
+          Rails.application.secrets.secret_key_base
+        )[0]
+        expect(decoded_token).to be
+        %w(id service_id exp).each do |expected_key|
+          expect(decoded_token).to have_key(expected_key)
+        end
+        expect(decoded_token['id']).to eq(existing_user.id)
+        expect(decoded_token['service_id']).to eq(authentication_service.service_id)
+        existing_user = User.find(decoded_token['id'])
+        expect(existing_user).to be
+        expect(existing_user.id).to eq(existing_user.id)
+        expect(existing_user.last_login_at.to_i).not_to eq(original_last_login_at)
+        expect(existing_user.last_login_at.to_i).to be >= pre_time
+        expect(existing_user.last_login_at.to_i).to be <= post_time
+      end
+
+      context 'nil payload' do
+        let(:payload) { nil }
+        it 'should respond with an error when not provided an access_token' do
+          is_expected.to eq(400)
+          expect(response.status).to eq(400)
+          expect(response.body).to be
+          error_response = JSON.parse(response.body)
+          %w(error reason suggestion).each do |expected_key|
+            expect(error_response).to have_key expected_key
+          end
+          expect(error_response['error']).to eq(400)
+          expect(error_response['reason']).to eq('no access_token')
+          expect(error_response['suggestion']).to eq('you might need to login through an authentication service')
+        end
+      end
+
+      context 'nil access_token' do
+        let(:payload) {
+          if authentication_service.is_default?
```

Comment:
An `if` statement in a test is a code smell. Let's look through this together and see what we can break off into its own shared example and refactor using specific contexts for default and non-default auth services.